InteractiveUtils: support type annotations as substitutes for values

serenity4 opened the pull request:
This PR extends code introspection macros (`@which`, `@code_typed` and friends) to recognize type annotations of the form `f(1, ::Float64)` as types to be forwarded as is to the relevant function. Here are a few examples:
```julia
@code_typed f(1, ::Float64)
@code_typed f(1, ::Any, 3, ::String...)
@code_typed f(1, ::Any, 3, ::Vararg{String}) # equivalent to the above
@code_typed ref.x = ::Float64
@code_typed (::Task).scope = ::Any
@code_typed ref[] = ::Float64
@code_typed (::Task).result
@code_typed (::Vector{Int})[::Int] = ::Float64
@code_typed (::typeof(sum))(1, 2) # only works for singleton function types at the moment
@code_typed [2 3 (::Int)] # mind the parentheses, `x ::T` parses as `x::T`
@code_typed sum(::Vector{T}; init = ::T) where {T<:Real}
```

This has a few advantages:
- It enables queries with non-concrete argument types, such as `@code_typed f(::Any)` which will be destructured as `code_typed(f, (Any,))`. Previously, macros could not express this and users had to rely on using the functional form.
- `Vararg` arguments may be expressed as `f(::Float64...)` or `f(::Vararg{Float64})`, which was also not expressible previously.
- It is now possible to skip the construction of values. This is particularly useful for types difficult to construct on the spot, which may be the case when no straightforward constructor exists or when construction involves side effects, such as `Task`.

This is implemented quite straightforwardly by recursively parsing all `::T` annotations and turning them into an intermediate `InterpolatedType(T)` structure, which is unwrapped as `T` when the types of the arguments are extracted before calling `code_typed`.

I also added support for `(::typeof(sum))(vec)` forms, where a function type annotation is treated as its only instance (if we don't have a singleton type, an error is thrown), making it equivalent to `sum(vec)`. The original motivation was to implement type-annotating callable objects such as `(::Returns{Int})(args...)`, but as this requires changes to the underlying introspection functions I preferred to leave that out. I'd be happy to try adding support for it in a future PR if there are no objections.

serenity4 committed:
Handle Vararg in error hint

serenity4 committed:
Support type annotations in introspection macros

serenity4 committed:
Fix handling of Vararg annotations

serenity4 committed:
Update PR number

serenity4 committed:
Add NEWS entry

JeffBezanson commented:
Thank you, I've wanted this for a long time!

I don't really like the implementation method of using a special `InterpolatedType` type. It should be done by simply mapping the input expression to an output expression, where `::T` becomes `T` and anything else `x` becomes `Typeof(x)`.

JeffBezanson commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -37,20 +50,59 @@ function recursive_dotcalls!(ex, args, i=1)
+            T = subex.args[1].args[1]
+            ex.args[i] = :(Vararg{$T})
+        else
+            process_type_annotations!(subex)
```
Does this need to be recursive? I would think we can only handle `::T` as a direct argument to the function being examined.

serenity4 replied:
I initially thought we may want to support tuples/named tuples recursively, such as `@code_typed f(x, (1, (::Float64, 2)))` (where a similar motivation had me introduce `InterpolatedType` to support runtime splatting such as `f(1, (1, ::Float64)...)`.

But on second thought I'd rather keep user expectations simple with regards to the processing it does, so let's make it only support direct arguments.

topolarity commented:
Does this also support `UnionAll` signatures like `@code_typed accumulate(::Vector{T}, ::T, z) where T<:Integer`?

serenity4 commented:
>  Does this also support `UnionAll` signatures like `@code_typed accumulate(::Vector{T}, ::T, z) where T<:Integer`?

It doesn't support matching the same `TypeVar` to several type parameters, currently only  `@code_typed accumulate(::Vector{<:Integer}, ::Integer, z)` would work. But that's a great idea, and it shouldn't be hard actually if we simply extract any top-level `where` clauses and feed them alongside the final `Tuple{...}` type.

topolarity commented:
Not to keep tacking on feature requests, but `Varargs` support would also be very nice (e.g. `@code_typed +(::Int...)`)

serenity4 commented:
That's already there :)

serenity4 committed:
Directly emit types, don't recurse for type extraction

serenity4 committed:
Remove type annotation support for macros

It is not worth the possible interference with macro-defined syntax for :(::) operands

serenity4 committed:
Support `where` clauses

serenity4 commented:
> I don't really like the implementation method of using a special `InterpolatedType` type.

Agreed, especially that it is leaked to the user in error messages when something goes wrong. I removed it.

serenity4 committed:
Add more complex example to NEWS.md

serenity4 committed:
Fix/reimplement Test.inferred macro using `gen_call_with_extracted_types`

This macro broke in a few cases, and while fixing it it became clear
that the macro could simply wrap around `gen_call_with_extracted_types`

serenity4 committed:
Re-add support for argument splatting

serenity4 committed:
Re-add support for splatting parameters

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into introspection-type-args

serenity4 committed:
Minor fixes

serenity4 committed:
Revert refactor of Test.inferred macro

This caused more trouble than it is worth, best to
keep the current implementation as is for now.

serenity4 committed:
Fix doctest failure

serenity4 committed:
Misc improvements

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into introspection-type-args

topolarity reviewed (commented):
Didn't do a complete read-through yet, but it's looking pretty good to me - Thanks @serenity4 !

serenity4 commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -37,20 +57,85 @@ function recursive_dotcalls!(ex, args, i=1)
+typeof_nt(nt::NamedTuple) = typeof(nt)
+typeof_nt(nt::Base.Pairs) = typeof(values(nt))
+
+function namedtuple_type(@nospecialize(nt1::Type{<:NamedTuple}), @nospecialize(nt2::Type{<:NamedTuple}))
```
Ah, I thought we would just error in that case and even made a test for it :sweat_smile:
```julia
kwargs_1 = (; digits = 3)
@test_throws "is not unique" @which round(1.2; digits = 1, kwargs_1...)
```

I'll add support for that then.

aviatesk commented on base/errorshow.jl:
```diff
@@ -1072,7 +1072,7 @@ Experimental.register_error_hint(nonsetable_type_hint_handler, MethodError)
 function string_concatenation_hint_handler(io, ex, arg_types, kwargs)
     @nospecialize
-    if (ex.f === +) && !isempty(arg_types) && all(i -> i <: AbstractString, arg_types)
+    if (ex.f === +) && !isempty(arg_types) && all(i -> unwrapva(i) <: AbstractString, arg_types)
```
```suggestion
    if (ex.f === +) && !isempty(arg_types) && all(@nospecialize(a) -> unwrapva(a) <: AbstractString, arg_types)
```

aviatesk commented on base/errorshow.jl:
```diff
@@ -1072,7 +1072,7 @@ Experimental.register_error_hint(nonsetable_type_hint_handler, MethodError)
 # (probably attempting concatenation)
 function string_concatenation_hint_handler(io, ex, arg_types, kwargs)
     @nospecialize
```
```suggestion
function string_concatenation_hint_handler(@nospecialize(io::IO), ex::MethodError, arg_types::Vector{Any}, kwargs::Vector{Any})
```

aviatesk commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -37,20 +57,85 @@ function recursive_dotcalls!(ex, args, i=1)
     return ex, i
 end
 
+function extract_farg(arg)
```
```suggestion
function extract_farg(@nospecialize arg)
```

aviatesk commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -37,20 +57,85 @@ function recursive_dotcalls!(ex, args, i=1)
+    args, kwargs
+end
+
+function namedtuple_type(kwargs)
```
Or we can even restrict it to `kwargs::Vector{Any}`.
Since these internal helper functions aren't really expected to be overloaded, it's better for maintenance to declare the methods using concrete types whenever possible.

aviatesk commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -37,20 +57,85 @@ function recursive_dotcalls!(ex, args, i=1)
     end
     if isa(ex0, Expr)
-        if ex0.head === :do && Meta.isexpr(get(ex0.args, 1, nothing), :call)
+        ex0, where_params = extract_where_parameters(ex0)
+        if ex0.head === :do && isexpr(get(ex0.args, 1, nothing), :call)
```
Probably better add guard against `!(ex0::Expr)` possibility, or add type assertion in `extract_where_parameters`.

aviatesk commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -37,20 +57,85 @@ function recursive_dotcalls!(ex, args, i=1)
                 return Expr(:call, :error, "ill-formed do call")
             end
-            i = findlast(a->(Meta.isexpr(a, :kw) || Meta.isexpr(a, :parameters)), ex0.args[1].args)
+            i = findlast(a->(isexpr(a, :kw) || isexpr(a, :parameters)), ex0.args[1].args)
```
```suggestion
            i = findlast(@nospecialize(a)->(isexpr(a, :kw) || isexpr(a, :parameters)), ex0.args[1].args)
```

aviatesk commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -102,32 +187,35 @@ function gen_call_with_extracted_types(__module__, fcn, ex0, kws=Expr[])
-                $(fcn)(Core.kwcall,
-                       Tuple{typeof(kwargs), Core.Typeof(arg1), map(Core.Typeof, args)...};
-                       $(kws...))
+        if any(a->(isexpr(a, :kw) || isexpr(a, :parameters)), ex0.args)
```
```suggestion
        if any(@nospecialize(a)->(isexpr(a, :kw) || isexpr(a, :parameters)), ex0.args)
```

aviatesk commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -2,14 +2,34 @@
+    Expr(:where, ex, esc.(where_params)...)
+end
+
+function get_typeof(ex)
```
```suggestion
function get_typeof(@nospecialize ex)
```

aviatesk commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -170,16 +254,14 @@ function gen_call_with_extracted_types(__module__, fcn, ex0, kws=Expr[])
     exret = Expr(:none)
     if ex.head === :call
-        if any(e->(isa(e, Expr) && e.head === :(...)), ex0.args) &&
+        if any(x -> isexpr(x, :...), ex0.args) &&
```
```suggestion
        if any(@nospecialize(x) -> isexpr(x, :...), ex0.args) &&
```

aviatesk commented:
I added  a few minor comments on the code style, but I think the fundamental approach is correct. Great job implementing this feature!

serenity4 committed:
Apply suggestions from code review

Co-authored-by: Shuhei Kadowaki <40514306+aviatesk@users.noreply.github.com>

serenity4 committed:
Add more type annotations to function arguments

serenity4 committed:
Ensure `ex0` remains an `Expr` after `where` parameter extraction

serenity4 committed:
Add missing nospecialize

serenity4 committed:
Account for duplicates when merging parameter types

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into introspection-type-args

serenity4 commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -102,32 +187,35 @@ function gen_call_with_extracted_types(__module__, fcn, ex0, kws=Expr[])
+            return :($(fcn)(Core.kwcall, $tt; $(kws...)))
         elseif ex0.head === :call
+            argtypes = Any[get_typeof(arg) for arg in ex0.args[2:end]]
             if ex0.args[1] === :^ && length(ex0.args) >= 3 && isa(ex0.args[3], Int)
```
Yeah... don't look too closely at this ancient code, you may find other weird things :sweat_smile:
